### PR TITLE
T5 enable self-attention kv caching

### PR DESCRIPTION
**Pull request by kundaMwiza**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by kundaMwiza** [comment]
test script
```
import torch
from datasets import load_dataset
import transformers
from transformers import AutoTokenizer, default_data_collator, AutoModelForSeq2SeqLM
from torch.utils.data import DataLoader
from datasets import load_dataset
from optimum.graphcore.modeling_utils import to_pipelined
from optimum.graphcore import IPUConfig

model_name_or_path = "t5-small"
encoder_max_length = 128
decoder_max_length = 128
batch_size = 8

source_lang = "en"
target_lang = "hu"
dataset = load_dataset("opus_books", f"{source_lang}-{target_lang}")
dataset["train"] = dataset["train"].select(range(10000))

# data preprocessing
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
prefix = "translate English to Hungarian: "
def preprocess_function(examples):
    inputs = [prefix + seq[source_lang] for seq in examples["translation"]]
    targets = [seq[target_lang] for seq in examples["translation"]]
    model_inputs = tokenizer(inputs, max_length=encoder_max_length, padding="max_length", truncation=True, return_tensors="pt")
    labels = tokenizer(targets, max_length=decoder_max_length, padding="max_length", truncation=True, return_tensors="pt")
    labels = labels["input_ids"]
    labels[labels == tokenizer.pad_token_id] = -100
    model_inputs["labels"] = labels
    return model_inputs

processed_datasets = dataset.map(
    preprocess_function,
    batched=True,
    remove_columns=dataset["train"].column_names,
    load_from_cache_file=False,
    desc="Running tokenizer on dataset",
)

train_dataset = processed_datasets["train"]

train_dataloader = DataLoader(
    train_dataset, shuffle=False, collate_fn=default_data_collator, batch_size=batch_size, pin_memory=True
)

dloader = iter(train_dataloader)
batch = next(dloader)
batch.pop("labels")

ipu_config = IPUConfig(layers_per_ipu=[-1, -1], ipus_per_replica=2).eval()

use_cache=True
model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
model = to_pipelined(model, ipu_config).half().parallelize(for_generation=True, use_cache=use_cache, batch_size=batch_size, max_length=decoder_max_length, num_beams=1)
for block in model.decoder.block:
    assert block.layer[0].SelfAttention.kv_cache_initialized  == use_cache
cache_output = model.generate(**batch, use_cache=use_cache, max_length=decoder_max_length)
model.destroy()

model = model.deparallelize().float()
model.__class__ = transformers.models.t5.modeling_t5.T5ForConditionalGeneration
cpu_cache_output = model.generate(**batch, use_cache=use_cache, max_length=decoder_max_length)
torch.testing.assert_close(cache_output, cpu_cache_output, rtol=0, atol=0)

use_cache=False
model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
model = to_pipelined(model, ipu_config).half().parallelize(for_generation=True, use_cache=use_cache, batch_size=batch_size, max_length=decoder_max_length, num_beams=1)
for block in model.decoder.block:
    assert block.layer[0].SelfAttention.kv_cache_initialized  == use_cache
output = model.generate(**batch, use_cache=use_cache, max_length=decoder_max_length)

model = model.deparallelize().float()
model.__class__ = transformers.models.t5.modeling_t5.T5ForConditionalGeneration
cpu_output = model.generate(**batch, use_cache=use_cache, max_length=decoder_max_length)
torch.testing.assert_close(output, cpu_output, rtol=0, atol=0)
```

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -62,6 +63,137 @@ def forward(self, input: Tensor) -> Tensor:
...
+        attn_output = self.o(attn_output)
+
+        # TODO: disable return present_key_value_state since caching is on device?
+        present_key_value_state = (key_states, value_states) if (self.is_decoder and use_cache) else None
+        outputs = (attn_output,) + (present_key_value_state,) + (position_bias,)
+
+        if output_attentions:
+            outputs = outputs + (attn_weights,)
+
+        if hasattr(self, "_step"):
```
Comment (katalinic-gc) [inline optimum/graphcore/models/t5/modeling_t5.py:191]: I don't believe this is used

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -62,6 +63,137 @@ def forward(self, input: Tensor) -> Tensor:
...
+        )  # (batch_size, n_heads, seq_length, key_length)
+
+        # Mask heads if we want to
+        if layer_head_mask is not None:
+            attn_weights = attn_weights * layer_head_mask
+
+        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self.o(attn_output)
+
+        # TODO: disable return present_key_value_state since caching is on device?
```
Comment (katalinic-gc) [inline optimum/graphcore/models/t5/modeling_t5.py:184]: For generation I'd expect the generation wrappers we have to take care of returning only relevant values

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -174,7 +306,277 @@ def invert_attention_mask(self, *args, **kwargs) -> Tensor:
...
+        inputs_embeds=None,
+        head_mask=None,
+        cross_attn_head_mask=None,
+        past_key_values=None,
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        """The same as super().forward but the attention mask is only internally generated if `use_cache` is False.
```
Comment (katalinic-gc) [inline optimum/graphcore/models/t5/modeling_t5.py:370]: To patch as discussed, intercept to create a mask with `update_attention_mask`.

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -62,6 +63,137 @@ def forward(self, input: Tensor) -> Tensor:
...
+            # the prefix tokens
+            if key_value_states is None:
+                # caching key states for self attention
+                # self-attn
+                # (batch_size, n_heads, key_length, dim_per_head)
+                tgt_len = key_states.shape[-2]
+                if tgt_len != 1:
+                    raise ValueError(f"KV cache expects tgt_len = 1, received {tgt_len}.")
+                key_states, value_states = self.add_to_kv_cache(key_states, value_states)
+                mask = self.update_attention_mask(mask)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/t5/modeling_t5.py:130]: Just marking this for removal as you pointed out

File: optimum/graphcore/generation/attention_mixin.py [hunk 5]
```diff
@@ -40,6 +40,7 @@ class IPUAttentionMixin:
 
     _kv_cache_initialized: bool = False
     _cross_kv_cache_initialized: bool = False
+    _kv_cache_max_length: int = 1
```
Comment (katalinic-gc) [inline optimum/graphcore/generation/attention_mixin.py:43]: Please remove this, the value is only relevant when use_cache is True at which point you can just refer to `self._k_cache.shape[-2]` in the property.

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 6]
```diff
@@ -62,6 +63,137 @@ def forward(self, input: Tensor) -> Tensor:
...
+                if self.gradient_checkpointing and self.training:
+                    position_bias.requires_grad = True
+            else:
+                position_bias = self.compute_bias(real_seq_length, key_length, device=scores.device)
+
+            # if key and values are already calculated
+            # we want only the last query position bias
+            # if kv_cache init or cross_kv_cache init
+            if self.kv_cache_initialized:
+                position_bias = poptorch.dynamic_slice(position_bias, 2, self.generation_step, 1, 1)
```
Comment (katalinic-gc) [inline optimum/graphcore/models/t5/modeling_t5.py:155]: Could you change this to `_generation_step` as is currently done throughout the library, and if desired we can look at changing all references to `generation_step` in a separate PR.

**Review by katalinic-gc — APPROVED** [review]
